### PR TITLE
Fixes #379 #434

### LINKED ISSUE #379: Types | StyledSubcomponentType is too rigid and complains about regular JSX
The intent behind the architecture is to allow any JSX to be passed into those `Styled_________` props. While we encourage restyling our internal base components with styled-components - we shouldn't complain if the developer doesn't do that, as flexibility is the selling-point of Foundry UI.

In the StyledSubcomponentType type, I would try to union the current type with `JSX.Element`.

### PR DESCRIPTION
**Pull request by xavierdonnellon**
**This PR accomplishes the following:** 
- Adds type support for custom component builders with props for columns in the `Columns` type, for things like cellComponent, headerCellComponent, etc, while retaining support for StyledComponents (see `ComponentBuilder`)
- Fixes group headers appearance in responsive view
- Makes custom sortFunctions functional in both directions, and extracts reusable comparison code into the `compareEntries` function, fixes group sorting
- Fixes sorting on the Table storyboard
- Fixes responsive view 
- Fixes passing in of rowProps StyledRow for non-grouped Table's
- Allows for styling of the ResponsiveHeaderCell, and wraps cell in CellContainer (which contains both the ResponsiveHeaderCell if in responsive mode, and the Cell).
- Change type of table `data` to `RowEntry[]`. Originally was `ColumnTypes`, which has no overlap with expected input. More clear typing.

### CODE REVIEW

File: src/components/Table/types.ts [hunk 1]
```diff
@@ -11,30 +11,45 @@ export type InternalExpansionIconProps = {
...
+  footerContent?: ReactNode;
+  sortFunction?: (item1: any, item2: any) => boolean;
+  isGroupLabel?: boolean;
+  cellComponent?: ComponentBuilder;
+  rowComponent?: ComponentBuilder;
+  headerCellComponent?: ComponentBuilder;
+  groupCellComponent?: ComponentBuilder;
 }
 
+type ComponentBuilder = StyledSubcomponentType | ((props: any) => JSX.Element) | React.FC;
```
Comment (aVileBroker) [inline src/components/Table/types.ts:32]: Didn't we move this to a separately exported type?

File: src/components/Table/types.ts [hunk 2]
```diff
@@ -11,30 +11,45 @@ export type InternalExpansionIconProps = {
...
+
+/**
+ * @deprecated Use `ColumnTypes`.
```
Comment (aVileBroker) [inline src/components/Table/types.ts:35]:
```suggestion
 * @deprecated Use `Columns`.
```

File: src/components/Table/types.ts [hunk 3]
```diff
@@ -11,30 +11,45 @@ export type InternalExpansionIconProps = {
...
+/** The data for a single row. */
+export type RowEntry = Record<string, any>;
+
 export type TableProps = {
   areGroupsCollapsible?: boolean;
   columnGap?: string;
-  columns: columnTypes;
-  data?: columnTypes[] | Array<Array<columnTypes>>;
+  columns: Columns;
+  data?: RowEntry[] | Array<Array<RowEntry>>;
```
Comment (aVileBroker) [inline src/components/Table/types.ts:46]:
```suggestion
  data?: Array<RowEntry> | Array<Array<RowEntry>>;
```

or 

```suggestion
  data?: Rows | Array<Rows>;
```

File: src/components/Table/types.ts [hunk 4]
```diff
@@ -43,34 +58,53 @@ export type TableProps = {
   StyledGroupLabelRow?: StyledSubcomponentType;
   StyledHeader?: StyledSubcomponentType;
   StyledHeaderCell?: StyledSubcomponentType;
+  /**
+   * If in responsive mode (width < `minWidthBreakpoint`), the `StyledResponsiveHeaderCell` will appear next to the
+   * `StyledCell`. By default, it displays the `name` of the column. This component is not displayed if not in responsive
+   * mode, or if the column has no `name`.
+   */
```
Comment (aVileBroker) [inline src/components/Table/types.ts:65]:
```suggestion
  /**
   * Header cells used when the Table is in condensed view
   * based on `minWidthBreakpoint`
   */
```

File: src/components/Table/types.ts [hunk 5]
```diff
@@ -43,34 +58,53 @@ export type TableProps = {
...
+  StyledResponsiveHeaderCell?: StyledSubcomponentType;
   StyledRow?: StyledSubcomponentType;
   StyledFooter?: StyledSubcomponentType;
   StyledFooterCell?: StyledSubcomponentType;
+  /**
+   * If in responsive mode (width < `minWidthBreakpoint`), the `CellContainer` contains the `StyledCell`
+   * and the `StyledResponsiveHeaderCellA `CellContainer` contains the `StyledCell`, as well as the `StyledResponsiveHeaderCell` for each cell
+   * if width is less than `minWidthBreakpoint`. If width is greater than `minWidthBreakpoint`, then the `CellContainer`
+   * only wraps the `StyledCell`.
+   */
```
Comment (aVileBroker) [inline src/components/Table/types.ts:75]:
```suggestion
  /**
   * Extra wrapper around cells, which also contains the ResponsiveHeaderCell when in the condensed view
   */
```

File: src/components/Table/types.ts [hunk 6]
```diff
@@ -43,34 +58,53 @@ export type TableProps = {
...
   columnGap: string;
   columnWidths: string;
   rowNum?: number;
   reachedMinWidth?: boolean;
   isCollapsed?: boolean;
 };
+
 export type CellOptions = {
-  RenderedCell: any;
+  RenderedCell: ComponentBuilder;
```
Comment (aVileBroker) [inline src/components/Table/types.ts:104]: Should be StyledSubComponent?

**Review by aVileBroker — CHANGES_REQUESTED** [review]

**Review by aVileBroker — DISMISSED** [review]

File: src/components/Table/Table.tsx [hunk 7]
```diff
@@ -176,7 +176,7 @@ export const Cell = styled(StyledBaseTD)`
   word-break: break-word;
   hyphens: auto;
   width: unset;
-  padding: 0.5em;
+  padding: 0.5em 0;
```
Comment (aVileBroker) [inline src/components/Table/Table.tsx:179]: Why this change?

File: src/components/Table/types.ts (hunk 1 above)
Comment (xavierdonnellon) [reply src/components/Table/types.ts]: Re-added to StyledSubcomponentType

File: src/components/Table/Table.tsx (hunk 7 above)
Comment (xavierdonnellon) [reply src/components/Table/Table.tsx:179]: No need for horizontal padding on cells (makes it not line up with header)

Comment (xavierdonnellon) [reply src/components/Table/Table.tsx:179]: Also, half the padding comes from the Cell itself (the space between the text and the CellContainer) and half the padding went into the CellContainer (space between two rows)

**Review by aVileBroker — APPROVED** [review]
